File: pm/agvmgr/worker/mqtt_worker.py
#!/usr/bin/env python3
"""
agvmgr MQTT worker – VDA 5050 v2.0 pozíció rögzítő + Omron továbbítás
"""

import json
import math
import logging
import sys
import signal
import time
from datetime import datetime, timezone

import mysql.connector
import paho.mqtt.client as mqtt

# ── Konfiguráció ─────────────────────────────────────────────────────────────
DB_CONFIG = {
    'host':       'localhost',
    'user':       'robot',
    'password':   'abrakadabra',
    'database':   'agvmgr',
    'charset':    'utf8mb4',
    'autocommit': True,
}
LOG_FILE = '/var/log/agvmgr_worker.log'

# ── Logging ───────────────────────────────────────────────────────────────────
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s',
    handlers=[
        logging.FileHandler(LOG_FILE),
        logging.StreamHandler(sys.stdout),
    ]
)
log = logging.getLogger(__name__)

# ── Globális állapot ──────────────────────────────────────────────────────────
_db_conn      = None
topic_map     = {}   # full_topic -> agv_id
agv_meta      = {}   # agv_id -> {'name': ..., 'serial_no': ...}
omron_fwd     = {}   # agv_id -> {'topic': ..., 'fields': [...], 'enabled': bool}
omron_client  = None


# ── DB kapcsolat ──────────────────────────────────────────────────────────────
def db_get():
    global _db_conn
    try:
        if _db_conn and _db_conn.is_connected():
            return _db_conn
    except Exception:
        pass
    _db_conn = mysql.connector.connect(**DB_CONFIG)
    log.info("DB kapcsolat megnyitva")
    return _db_conn


def db_query(sql, params=None, fetchone=False):
    conn = db_get()
    cur  = conn.cursor(dictionary=True)
    cur.execute(sql, params or ())
    result = cur.fetchone() if fetchone else cur.fetchall()
    cur.close()
    return result


def db_execute(sql, params=None):
    conn = db_get()
    cur  = conn.cursor()
    cur.execute(sql, params or ())
    cur.close()


# ── Konfig betöltés ───────────────────────────────────────────────────────────
def load_agvs():
    global topic_map, agv_meta
    rows = db_query("SELECT id, name, serial_no, topic FROM agv WHERE enabled = 1")
    topic_map = {}
    agv_meta  = {}
    for row in rows:
        base = row['topic'].rstrip('/')
        topic_map[base + '/visualization'] = row['id']
        topic_map[base + '/state']         = row['id']
        agv_meta[row['id']] = {'name': row['name'] or row['serial_no'], 'serial_no': row['serial_no']}
    log.info(f"Betöltve {len(rows)} AGV, {len(topic_map)} topic figyelve")
    return list(topic_map.keys())


def load_omron_fwd():
    global omron_fwd
    rows = db_query("""
        SELECT f.agv_id, f.topic_template, f.fields, f.enabled
        FROM omron_forward f
        JOIN agv a ON a.id = f.agv_id
        WHERE a.enabled = 1
    """)
    omron_fwd = {}
    for row in rows:
        fields = json.loads(row['fields']) if row['fields'] else []
        omron_fwd[row['agv_id']] = {
            'topic':   row['topic_template'],
            'fields':  fields,
            'enabled': bool(row['enabled']),
        }
    log.info(f"Omron forward konfig betöltve: {len(omron_fwd)} AGV")


def get_broker_config(table='mqtt_broker'):
    return db_query(f"SELECT ip, port, username, password FROM {table} WHERE id=1", fetchone=True)


# ── Payload mentése DB-be ─────────────────────────────────────────────────────
def save_position(agv_id: int, payload_raw: str, source: str):
    try:
        payload = json.loads(payload_raw)
    except json.JSONDecodeError as e:
        log.warning(f"JSON parse hiba (agv {agv_id}): {e}")
        return None

    pos = payload.get('agvPosition') or {}
    vel = payload.get('velocity')    or {}
    bat = payload.get('batteryState') or {}

    x          = pos.get('x')
    y          = pos.get('y')
    theta      = pos.get('theta')
    map_id     = pos.get('mapId', '') or ''
    pos_init   = pos.get('positionInitialized')
    loc_score  = pos.get('localizationScore')
    dev_range  = pos.get('deviationRange')
    vx         = vel.get('vx')
    vy         = vel.get('vy')
    omega      = vel.get('omega')
    bat_charge = bat.get('batteryCharge')
    bat_volt   = bat.get('batteryVoltage')
    op_mode    = payload.get('operatingMode') or ''
    driving    = payload.get('driving')
    paused     = payload.get('paused')

    speed_str = ''
    if vx is not None and vy is not None:
        speed_str = f"  v={math.sqrt(vx**2+vy**2):.2f}m/s"
    deg_str = f"{math.degrees(theta):.1f}°" if theta is not None else '–'
    log.info(f"AGV {agv_id} [{source}]  x={x} y={y} θ={deg_str}{speed_str}  bat={bat_charge}%  mode={op_mode or '–'}")

    try:
        db_execute("""
            INSERT INTO agv_coords
              (agv_id, x, y, theta, map_id, position_initialized,
               localization_score, deviation_range,
               vx, vy, omega,
               battery_charge, battery_voltage,
               operating_mode, driving, paused,
               source, raw_payload, updated_at)
            VALUES
              (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,NOW(3))
            ON DUPLICATE KEY UPDATE
              x                    = COALESCE(%s, x),
              y                    = COALESCE(%s, y),
              theta                = COALESCE(%s, theta),
              map_id               = IF(%s <> '', %s, map_id),
              position_initialized = COALESCE(%s, position_initialized),
              localization_score   = COALESCE(%s, localization_score),
              deviation_range      = COALESCE(%s, deviation_range),
              vx                   = COALESCE(%s, vx),
              vy                   = COALESCE(%s, vy),
              omega                = COALESCE(%s, omega),
              battery_charge       = COALESCE(%s, battery_charge),
              battery_voltage      = COALESCE(%s, battery_voltage),
              operating_mode       = IF(%s <> '', %s, operating_mode),
              driving              = COALESCE(%s, driving),
              paused               = COALESCE(%s, paused),
              source               = %s,
              raw_payload          = %s,
              updated_at           = NOW(3)
        """, (
            agv_id, x, y, theta, map_id, pos_init,
            loc_score, dev_range, vx, vy, omega,
            bat_charge, bat_volt, op_mode, driving, paused,
            source, payload_raw,
            x, y, theta,
            map_id, map_id,
            pos_init, loc_score, dev_range,
            vx, vy, omega,
            bat_charge, bat_volt,
            op_mode, op_mode,
            driving, paused,
            source, payload_raw,
        ))
    except Exception as e:
        log.error(f"DB mentési hiba (agv {agv_id}): {e}")
        global _db_conn
        _db_conn = None

    # Visszaadjuk a parsolt adatokat az Omron forwarding-hoz
    return {
        'x': x, 'y': y, 'theta': theta,
        'theta_deg': round(math.degrees(theta), 4) if theta is not None else None,
        'map_id': map_id or None,
        'pos_init': pos_init,
        'loc_score': loc_score,
        'dev_range': dev_range,
        'vx': vx, 'vy': vy, 'omega': omega,
        'speed': round(math.sqrt(vx**2+vy**2), 4) if (vx is not None and vy is not None) else None,
        'battery': bat_charge,
        'voltage': bat_volt,
        'mode': op_mode or None,
        'driving': driving,
        'paused': paused,
    }


# ── Omron forwarding ──────────────────────────────────────────────────────────
def forward_to_omron(agv_id: int, parsed: dict):
    global omron_client
    if omron_client is None:
        return

    cfg = omron_fwd.get(agv_id)
    if not cfg or not cfg['enabled'] or not cfg['fields']:
        return

    meta = agv_meta.get(agv_id, {})
    topic = cfg['topic'] \
        .replace('{serial_no}', meta.get('serial_no', str(agv_id))) \
        .replace('{name}',      meta.get('name', str(agv_id)))

    out = {}
    field_map = {
        'x':         ('x',         parsed.get('x')),
        'y':         ('y',         parsed.get('y')),
        'theta':     ('theta',     parsed.get('theta')),
        'theta_deg': ('theta_deg', parsed.get('theta_deg')),
        'map_id':    ('map_id',    parsed.get('map_id')),
        'pos_init':  ('positionInitialized', parsed.get('pos_init')),
        'loc_score': ('localizationScore',   parsed.get('loc_score')),
        'dev_range': ('deviationRange',      parsed.get('dev_range')),
        'speed':     ('speed',     parsed.get('speed')),
        'vx':        ('vx',        parsed.get('vx')),
        'vy':        ('vy',        parsed.get('vy')),
        'omega':     ('omega',     parsed.get('omega')),
        'battery':   ('batteryCharge',   parsed.get('battery')),
        'voltage':   ('batteryVoltage',  parsed.get('voltage')),
        'mode':      ('operatingMode',   parsed.get('mode')),
        'driving':   ('driving',   parsed.get('driving')),
        'paused':    ('paused',    parsed.get('paused')),
        'timestamp': ('timestamp', datetime.now(timezone.utc).strftime('%Y-%m-%dT%H:%M:%S.%f')[:-3] + 'Z'),
        'agv_name':  ('agvName',   meta.get('name')),
        'serial_no': ('serialNo',  meta.get('serial_no')),
    }

    for key in cfg['fields']:
        if key in field_map:
            json_key, value = field_map[key]
            if value is not None:
                out[json_key] = value

    if not out:
        return

    try:
        omron_client.publish(topic, json.dumps(out), qos=1, retain=False)
        log.debug(f"Omron forward: {topic} → {list(out.keys())}")
    except Exception as e:
        log.warning(f"Omron publish hiba (agv {agv_id}): {e}")


# ── MQTT callbacks (AGV broker) ───────────────────────────────────────────────
def on_connect(client, userdata, flags, rc):
    codes = {0:'OK',1:'protokoll hiba',2:'érvénytelen ID',3:'szerver nem elérhető',4:'hibás jelszó',5:'nincs jog'}
    if rc == 0:
        log.info("AGV MQTT broker kapcsolódva")
        for topic in topic_map:
            client.subscribe(topic, qos=1)
            log.info(f"  → feliratkozva: {topic}")
    else:
        log.error(f"AGV MQTT kapcsolódás meghiúsult: {codes.get(rc, rc)}")


def on_message(client, userdata, msg):
    topic  = msg.topic
    agv_id = topic_map.get(topic)
    if agv_id is None:
        return
    source = 'visualization' if topic.endswith('/visualization') else 'state'
    payload_raw = msg.payload.decode('utf-8', errors='replace')
    parsed = save_position(agv_id, payload_raw, source)
    if parsed:
        forward_to_omron(agv_id, parsed)


def on_disconnect(client, userdata, rc):
    if rc != 0:
        log.warning(f"AGV MQTT kapcsolat megszakadt (rc={rc}), újracsatlakozás...")


# ── Omron MQTT kliens setup ───────────────────────────────────────────────────
def setup_omron_client():
    global omron_client
    omron_cfg = db_query("SELECT ip, port, username, password, enabled FROM omron_broker WHERE id=1", fetchone=True)

    if not omron_cfg or not omron_cfg.get('enabled') or not omron_cfg.get('ip'):
        if omron_cfg and omron_cfg.get('enabled') and not omron_cfg.get('ip'):
            log.warning("Omron forwarding engedélyezve, de nincs IP beállítva")
        else:
            log.info("Omron forwarding kikapcsolva")
        omron_client = None
        return

    c = mqtt.Client(client_id="agvmgr_omron_fwd", clean_session=True)
    if omron_cfg.get('username'):
        c.username_pw_set(omron_cfg['username'], omron_cfg.get('password') or '')

    c.reconnect_delay_set(min_delay=5, max_delay=60)

    def omron_on_connect(cl, ud, fl, rc):
        if rc == 0: log.info(f"Omron MQTT broker kapcsolódva ({omron_cfg['ip']}:{omron_cfg['port']})")
        else:       log.error(f"Omron MQTT kapcsolódás meghiúsult: rc={rc}")

    def omron_on_disconnect(cl, ud, rc):
        if rc != 0: log.warning(f"Omron MQTT kapcsolat megszakadt (rc={rc})")

    c.on_connect    = omron_on_connect
    c.on_disconnect = omron_on_disconnect

    try:
        c.connect(omron_cfg['ip'], int(omron_cfg['port']), keepalive=60)
        c.loop_start()
        omron_client = c
        log.info(f"Omron MQTT kliens elindítva → {omron_cfg['ip']}:{omron_cfg['port']}")
    except Exception as e:
        log.error(f"Omron MQTT csatlakozási hiba: {e}")
        omron_client = None


# ── Főprogram ─────────────────────────────────────────────────────────────────
def main():
    log.info("agvmgr MQTT worker indul")

    load_agvs()
    load_omron_fwd()

    broker = get_broker_config('mqtt_broker')
    if not broker or not broker.get('ip'):
        log.error("Nincs AGV MQTT broker IP beállítva. Kilépés.")
        sys.exit(1)

    setup_omron_client()

    client = mqtt.Client(client_id="agvmgr_worker", clean_session=True)
    client.on_connect    = on_connect
    client.on_message    = on_message
    client.on_disconnect = on_disconnect

    if broker.get('username'):
        client.username_pw_set(broker['username'], broker.get('password') or '')

    client.reconnect_delay_set(min_delay=2, max_delay=30)

    def shutdown(sig, frame):
        log.info("Worker leállítás (signal fogadva)")
        client.disconnect()
        if omron_client:
            omron_client.loop_stop()
            omron_client.disconnect()
        sys.exit(0)

    signal.signal(signal.SIGTERM, shutdown)
    signal.signal(signal.SIGINT,  shutdown)

    log.info(f"Csatlakozás AGV brokerhez: {broker['ip']}:{broker['port']}")
    try:
        client.connect(broker['ip'], int(broker['port']), keepalive=60)
    except Exception as e:
        log.error(f"AGV MQTT csatlakozási hiba: {e}")
        sys.exit(1)

    client.loop_forever(retry_first_connection=True)


if __name__ == '__main__':
    main()

File: pm/agvmgr/worker/mqtt_worker.php
#!/usr/bin/env php
<?php
declare(strict_types=1);
set_time_limit(0);

define('DB_HOST',          'localhost');
define('DB_USER',          'robot');
define('DB_PASS',          'abrakadabra');
define('DB_NAME',          'agvmgr');
define('LOG_FILE',         '/var/log/agvmgr_worker.log');
define('CONF_TTL',         300);
define('RECONNECT_SEC',    5);
define('HISTORY_INTERVAL', 10);
define('OFFLINE_TIMEOUT',  90);
define('BAT_LOW',          20.0);
define('BAT_CRITICAL',     10.0);
define('BAT_OK',           25.0);

$pdo              = null;
$topic_map        = [];
$agv_meta         = [];
$running          = true;
$prev_state       = [];
$last_history     = [];
$last_offline_chk = 0;
$last_reload      = 0;

if (function_exists('pcntl_signal')) {
    $h = function () use (&$running): void { $running = false; };
    pcntl_signal(SIGTERM, $h);
    pcntl_signal(SIGINT,  $h);
}

function wlog(string $level, string $msg): void {
    $line = date('Y-m-d H:i:s') . " [$level] $msg\n";
    @file_put_contents(LOG_FILE, $line, FILE_APPEND | LOCK_EX);
    echo $line;
}

function db(): PDO {
    global $pdo;
    if ($pdo !== null) {
        try { $pdo->query('SELECT 1'); return $pdo; } catch (Exception $e) { $pdo = null; }
    }
    $pdo = new PDO(
        'mysql:host=' . DB_HOST . ';dbname=' . DB_NAME . ';charset=utf8mb4',
        DB_USER, DB_PASS,
        [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION, PDO::MYSQL_ATTR_FOUND_ROWS => true]
    );
    wlog('INFO', 'DB kapcsolat megnyitva');
    return $pdo;
}

function load_config(): array {
    global $topic_map, $agv_meta;
    $rows      = db()->query("SELECT id, name, serial_no, topic FROM agv WHERE enabled=1")->fetchAll(PDO::FETCH_ASSOC);
    $topic_map = [];
    $agv_meta  = [];
    foreach ($rows as $r) {
        $base = rtrim((string)$r['topic'], '/');
        $topic_map[$base . '/visualization'] = (int)$r['id'];
        $topic_map[$base . '/state']         = (int)$r['id'];
        $agv_meta[(int)$r['id']] = [
            'name'      => (string)($r['name'] ?: $r['serial_no']),
            'serial_no' => (string)$r['serial_no'],
        ];
    }
    wlog('INFO', 'Betöltve ' . count($rows) . ' AGV, ' . count($topic_map) . ' topic figyelve');

    $broker = db()->query("SELECT ip, port, username, password FROM mqtt_broker WHERE id=1 AND enabled=1")->fetch(PDO::FETCH_ASSOC);
    return $broker ?: [];
}

function build_command(array $broker): ?string {
    if (empty($broker['ip'])) {
        wlog('ERROR', 'Nincs broker IP beállítva.');
        return null;
    }
    $args = [
        'mosquitto_sub',
        '-h', $broker['ip'],
        '-p', (string)(int)($broker['port'] ?? 1883),
        '-t', '#',
        '-v',
        '--keepalive', '30',
    ];
    if (!empty($broker['username'])) {
        $args[] = '-u'; $args[] = $broker['username'];
        $args[] = '-P'; $args[] = $broker['password'] ?? '';
    }
    return implode(' ', array_map('escapeshellarg', $args));
}

function save_position(int $agv_id, string $raw, string $source): ?array {
    $p = json_decode($raw, true);
    if (!$p) { wlog('WARNING', "JSON parse hiba (agv $agv_id)"); return null; }

    $pos = $p['agvPosition']  ?? [];
    $vel = $p['velocity']     ?? [];
    $bat = $p['batteryState'] ?? [];

    $x         = isset($pos['x'])                   ? (float)$pos['x']                       : null;
    $y         = isset($pos['y'])                   ? (float)$pos['y']                       : null;
    $theta     = isset($pos['theta'])               ? (float)$pos['theta']                   : null;
    $map_id    = (string)($pos['mapId']             ?? '');
    $pos_init  = isset($pos['positionInitialized']) ? (int)(bool)$pos['positionInitialized'] : null;
    $loc_score = isset($pos['localizationScore'])   ? (float)$pos['localizationScore']       : null;
    $dev_range = isset($pos['deviationRange'])      ? (float)$pos['deviationRange']          : null;
    $vx        = isset($vel['vx'])                  ? (float)$vel['vx']                      : null;
    $vy        = isset($vel['vy'])                  ? (float)$vel['vy']                      : null;
    $omega     = isset($vel['omega'])               ? (float)$vel['omega']                   : null;
    $bat_chg   = isset($bat['batteryCharge'])       ? (float)$bat['batteryCharge']           : null;
    $bat_volt  = isset($bat['batteryVoltage'])      ? (float)$bat['batteryVoltage']          : null;
    $op_mode   = (string)($p['operatingMode']       ?? '');
    $driving   = isset($p['driving'])               ? (int)(bool)$p['driving']               : null;
    $paused    = isset($p['paused'])                ? (int)(bool)$p['paused']                : null;

    $deg = $theta !== null ? sprintf('%.1f°', rad2deg($theta)) : '–';
    $spd = ($vx !== null && $vy !== null) ? sprintf('  v=%.2fm/s', sqrt($vx ** 2 + $vy ** 2)) : '';
    wlog('INFO', "AGV $agv_id [$source]  x=$x y=$y θ=$deg$spd  bat={$bat_chg}%  mode=" . ($op_mode ?: '–'));

    try {
        $stmt = db()->prepare("
            INSERT INTO agv_coords
              (agv_id, x, y, theta, map_id, position_initialized, localization_score,
               deviation_range, vx, vy, omega, battery_charge, battery_voltage,
               operating_mode, driving, paused, source, raw_payload, updated_at)
            VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,NOW(3))
            ON DUPLICATE KEY UPDATE
              x                    = COALESCE(?,x),
              y                    = COALESCE(?,y),
              theta                = COALESCE(?,theta),
              map_id               = IF(?<>'',?,map_id),
              position_initialized = COALESCE(?,position_initialized),
              localization_score   = COALESCE(?,localization_score),
              deviation_range      = COALESCE(?,deviation_range),
              vx                   = COALESCE(?,vx),
              vy                   = COALESCE(?,vy),
              omega                = COALESCE(?,omega),
              battery_charge       = COALESCE(?,battery_charge),
              battery_voltage      = COALESCE(?,battery_voltage),
              operating_mode       = IF(?<>'',?,operating_mode),
              driving              = COALESCE(?,driving),
              paused               = COALESCE(?,paused),
              source               = ?,
              raw_payload          = ?,
              updated_at           = NOW(3)
        ");
        $stmt->execute([
            $agv_id, $x, $y, $theta, $map_id, $pos_init, $loc_score,
            $dev_range, $vx, $vy, $omega, $bat_chg, $bat_volt,
            $op_mode, $driving, $paused, $source, $raw,
            $x, $y, $theta, $map_id, $map_id, $pos_init, $loc_score, $dev_range,
            $vx, $vy, $omega, $bat_chg, $bat_volt, $op_mode, $op_mode,
            $driving, $paused, $source, $raw,
        ]);
    } catch (Exception $e) {
        wlog('ERROR', "DB mentési hiba (agv $agv_id): " . $e->getMessage());
        global $pdo; $pdo = null;
    }

    return [
        'x' => $x, 'y' => $y, 'theta' => $theta, 'map_id' => $map_id,
        'pos_init' => $pos_init, 'loc_score' => $loc_score, 'dev_range' => $dev_range,
        'vx' => $vx, 'vy' => $vy, 'omega' => $omega,
        'speed'    => ($vx !== null && $vy !== null) ? round(sqrt($vx ** 2 + $vy ** 2), 4) : null,
        'bat_chg'  => $bat_chg, 'bat_volt' => $bat_volt,
        'op_mode'  => $op_mode, 'driving' => $driving, 'paused' => $paused,
    ];
}

function detect_events(int $agv_id, array $p): void {
    global $prev_state, $agv_meta;
    $name = $agv_meta[$agv_id]['name'] ?? "AGV#$agv_id";
    $prev = $prev_state[$agv_id] ?? [
        'battery' => null, 'mode' => '', 'driving' => null,
        'paused'  => null, 'pos_init' => null, 'offline' => false,
    ];

    if (!empty($prev['offline'])) {
        log_event($agv_id, 'online', 'info', "$name visszatért online");
    }

    $bat = $p['bat_chg']; $pbat = $prev['battery'];
    if ($bat !== null && $pbat !== null) {
        if ($bat < BAT_CRITICAL && $pbat >= BAT_CRITICAL)
            log_event($agv_id, 'battery_critical', 'error',   "$name kritikus: " . round($bat, 1) . '%');
        elseif ($bat < BAT_LOW && $pbat >= BAT_LOW)
            log_event($agv_id, 'battery_low',      'warning', "$name alacsony: " . round($bat, 1) . '%');
        elseif ($bat >= BAT_OK && $pbat < BAT_OK)
            log_event($agv_id, 'battery_ok',       'info',    "$name feltöltve: " . round($bat, 1) . '%');
    }

    $mode = $p['op_mode']; $pmode = $prev['mode'];
    if ($mode && $pmode !== null && $mode !== $pmode && $pmode !== '') {
        log_event($agv_id, 'mode_change', 'info', "$name üzemmód: $pmode → $mode");
    }

    $drv = $p['driving']; $pdrv = $prev['driving'];
    if ($drv !== null && $pdrv !== null && $drv !== $pdrv) {
        log_event($agv_id, $drv ? 'driving_start' : 'driving_stop', 'info',
                  $drv ? "$name mozgás indult" : "$name megállt");
    }

    $pau = $p['paused']; $ppau = $prev['paused'];
    if ($pau !== null && $ppau !== null && $pau !== $ppau) {
        log_event($agv_id, $pau ? 'paused' : 'resumed', $pau ? 'warning' : 'info',
                  $pau ? "$name szünet" : "$name folytat");
    }

    $pi = $p['pos_init']; $ppi = $prev['pos_init'];
    if ($pi !== null && $ppi !== null && $pi !== $ppi) {
        log_event($agv_id, $pi ? 'pos_init' : 'pos_lost', $pi ? 'info' : 'warning',
                  $pi ? "$name pozíció init" : "$name pozíció elveszett");
    }

    $prev_state[$agv_id] = [
        'battery'  => $bat,
        'mode'     => $mode ?: $pmode,
        'driving'  => $drv  ?? $pdrv,
        'paused'   => $pau  ?? $ppau,
        'pos_init' => $pi   ?? $ppi,
        'offline'  => false,
    ];
}

function log_event(int $agv_id, string $type, string $severity, string $detail): void {
    try {
        db()->prepare("INSERT INTO agv_events (agv_id, event_type, severity, detail) VALUES (?,?,?,?)")
           ->execute([$agv_id, $type, $severity, $detail]);
        wlog('INFO', "Esemény [$severity] agv=$agv_id $type: $detail");
    } catch (Exception $e) {
        wlog('WARNING', "Esemény mentési hiba: " . $e->getMessage());
    }
}

function maybe_save_history(int $agv_id, array $p): void {
    global $last_history;
    $now = time();
    if (isset($last_history[$agv_id]) && $now - $last_history[$agv_id] < HISTORY_INTERVAL) return;
    $last_history[$agv_id] = $now;
    try {
        db()->prepare("INSERT INTO agv_coords_history (agv_id, x, y, theta, map_id, speed, battery, source) VALUES (?,?,?,?,?,?,?,?)")
           ->execute([$agv_id, $p['x'], $p['y'], $p['theta'], $p['map_id'] ?? '', $p['speed'], $p['bat_chg'], 'worker']);
    } catch (Exception $e) {
        wlog('WARNING', "History mentési hiba (agv $agv_id): " . $e->getMessage());
    }
}

function check_offline(): void {
    global $prev_state, $agv_meta;
    $rows = db()->query(
        "SELECT agv_id, TIMESTAMPDIFF(SECOND, updated_at, NOW()) AS age_sec
         FROM agv_coords WHERE TIMESTAMPDIFF(SECOND, updated_at, NOW()) > " . OFFLINE_TIMEOUT
    )->fetchAll(PDO::FETCH_ASSOC);
    foreach ($rows as $r) {
        $id = (int)$r['agv_id'];
        if (empty($prev_state[$id]['offline'])) {
            $name = $agv_meta[$id]['name'] ?? "AGV#$id";
            log_event($id, 'offline', 'warning', "$name offline – " . (int)$r['age_sec'] . " mp");
            $prev_state[$id]['offline'] = true;
        }
    }
}

function process_line(string $line): void {
    global $topic_map;
    if ($line === '') return;
    $sp = strpos($line, ' ');
    if ($sp === false) return;
    $topic   = substr($line, 0, $sp);
    $payload = substr($line, $sp + 1);
    $agv_id  = $topic_map[$topic] ?? null;
    if ($agv_id === null) return;
    $source = (substr($topic, -14) === '/visualization') ? 'visualization' : 'state';
    $parsed = save_position($agv_id, $payload, $source);
    if ($parsed) {
        detect_events($agv_id, $parsed);
        maybe_save_history($agv_id, $parsed);
    }
}

// ── Főprogram ─────────────────────────────────────────────────────────────────
wlog('INFO', 'agvmgr MQTT worker indul (mosquitto_sub alapú)');

$broker      = load_config();
$last_reload = time();

if (empty($broker['ip'])) {
    wlog('ERROR', 'Nincs AGV MQTT broker IP beállítva. Kilépés.');
    exit(1);
}

$proc  = null;
$pipes = [];

while ($running) {
    if (function_exists('pcntl_signal_dispatch')) pcntl_signal_dispatch();

    $now = time();

    // Konfig újratöltés
    if ($now - $last_reload >= CONF_TTL) {
        wlog('INFO', 'Konfig újratöltés...');
        $new_broker = load_config();
        if ($new_broker !== $broker) {
            wlog('INFO', 'Broker konfig megváltozott, újracsatlakozás...');
            $broker = $new_broker;
            if ($proc) { proc_terminate($proc); proc_close($proc); $proc = null; }
        }
        $last_reload = $now;
    }

    // Offline ellenőrzés 60 mp-enként
    if ($now - $last_offline_chk >= 60) {
        check_offline();
        $last_offline_chk = $now;
    }

    // Folyamat indítása / újraindítása
    if ($proc === null) {
        $cmd = build_command($broker);
        if ($cmd === null) { sleep(RECONNECT_SEC); continue; }

        wlog('INFO', "Csatlakozás: mosquitto_sub -h {$broker['ip']} -p {$broker['port']}");
        $desc = [1 => ['pipe', 'r'], 2 => ['pipe', 'r']];
        $proc = proc_open($cmd, $desc, $pipes);
        if (!is_resource($proc)) {
            wlog('ERROR', 'proc_open sikertelen, újrapróbálás ' . RECONNECT_SEC . ' mp múlva...');
            $proc = null;
            sleep(RECONNECT_SEC);
            continue;
        }
        stream_set_blocking($pipes[1], false);
        stream_set_blocking($pipes[2], false);
        wlog('INFO', 'mosquitto_sub folyamat elindítva');
    }

    // Státusz ellenőrzés
    $status = proc_get_status($proc);
    if (!$status['running']) {
        $stderr = stream_get_contents($pipes[2]);
        wlog('WARNING', 'mosquitto_sub kilépett (exit=' . $status['exitcode'] . ')'
            . ($stderr ? ': ' . trim($stderr) : ''));
        proc_close($proc);
        $proc = null;
        sleep(RECONNECT_SEC);
        continue;
    }

    // Sorok olvasása
    $line = fgets($pipes[1]);
    if ($line !== false) {
        process_line(trim($line));
    } else {
        usleep(10000); // 10 ms várakozás ha nincs adat
    }
}

wlog('INFO', 'Worker leállítás (signal fogadva)');
if ($proc) { proc_terminate($proc); proc_close($proc); }
